nbviewer/app.py

```diff
@@ -159,106 +172,138 @@ def init_tornado_application(self):
-                               'sections': frontpage_setup}
-    
-        # cache frontpage links for the maximum allowed time
-        max_cache_uris = {''}
-        for section in frontpage_setup['sections']:
-            for link in section['links']:
-                max_cache_uris.add('/' + link['target'])
-    
+        env.globals.update(nrhead=nrhead, nrfoot=nrfoot, git_data=git_data, jupyter_info=jupyter_info(), len=len)
+
+        return env
+
+    @property
+    def fetch_kwargs(self):
         fetch_kwargs = dict(connect_timeout=10,)
         if options.proxy_host:
-            fetch_kwargs.update(dict(proxy_host=options.proxy_host,
-                                     proxy_port=options.proxy_port))
-    
+            fetch_kwargs.update(proxy_host=options.proxy_host, proxy_port=options.proxy_port)
             log.app_log.info("Using web proxy {proxy_host}:{proxy_port}."
                              "".format(**fetch_kwargs))
-    
+        
         if options.no_check_certificate:
-            fetch_kwargs.update(dict(validate_cert=False))
-    
+            fetch_kwargs.update(validate_cert=False)
             log.app_log.info("Not validating SSL certificates")
-    
-        # prefer the jhub defined service prefix over the CLI
-        base_url = os.getenv('JUPYTERHUB_SERVICE_PREFIX', options.base_url)
+
+        return fetch_kwargs
+
+    @property
+    def formats(self):
+        formats = configure_formats(options, self.config, log.app_log)
+        return formats
+
+    # load frontpage sections
+    @property
+    def frontpage_setup(self):
+        with io.open(options.frontpage, 'r') as f:
+            frontpage_setup = json.load(f)
+        # check if the JSON has a 'sections' field, otherwise assume it is just a list of sessions,
+        # and provide the defaults of the other fields
+        if 'sections' not in frontpage_setup:
+            frontpage_setup = {
+                              'title':'nbviewer', 'subtitle':'A simple way to share Jupyter notebooks',
+                              'show_input':True, 'sections':frontpage_setup
+                              }
+        return frontpage_setup
+
+    @property
+    def pool(self):
+        if options.processes:
+            pool = ProcessPoolExecutor(options.processes)
+        else:
+            pool = ThreadPoolExecutor(options.threads)
+        return pool
+
+    @property
+    def rate_limiter(self):
+        rate_limiter = RateLimiter(limit=options.rate_limit, interval=options.rate_limit_interval, cache=self.cache)
+        return rate_limiter
+
+    @property
+    def template_paths(self):
+        template_paths = pjoin(here, 'templates')
+        if options.template_path is not None:
+            log.app_log.info("Using custom template path {}".format(options.template_path))
+            template_paths = [options.template_path, template_paths]
+
+        return template_paths
+
+
+    def init_tornado_application(self):
```

Comment:
I agree with your comment that it's hard to determine if all of the old command line parameters are having the same net effect after the refactor by reading the diff alone. I wonder if it would be possible to add a test that vets what's in the `settings` dict is what is expected (e.g., before and after these changes).

Reply:
Yes that sounds like it would be a good idea. 

I don't think I know how to do that. I don't really understand or know much about `unittests.mock` or `nosetests`. Any suggestions for what to write would be appreciated.

At the very least _everything_ has worked all of the (dozens maybe even hundreds) times I tested things "interactively" (for lack of a better word, i.e. running the program in the Docker container, as a standalone application and as a JupyterHub service, viewing all of the example notebooks, viewing GitHub users, repos, user gists, slides, etc). But I know this isn't enough to test the more advanced features like `threads` or `pools` or `memcache`. I'm not sure if I even have all of the hardware and/or software necessary to test those. I tried to ensure that I was just moving code verbatim as much as possible if I didn't fully understand it, and not inadvertently changing it. But the git diff is largely useless for verifying that, since it just registers things which are cut and pasted to another portion of the file as deleted.

Reply:
I'll need to take a look at the test suite to see how easily we can put something in. However, now that you reminded me about aliases, it might not be worth the effort.